File: src/main/kotlin/shortestpath/미래_도시.java
package shortestpath;

import java.util.*;

public class 미래_도시 {

    private static final int INF = (int)1e9;
    private static int[][] graph;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        //회사의 개수와 경로의 개수 입력받기
        int n = sc.nextInt();
        int m = sc.nextInt();
        //모든 최단경로 INF 로 초기화
        graph = new int[n+1][n+1];
        for(int i = 0; i <= n; i++){
            for(int j = 0; j <= n; j++){
                if(i == j) graph[i][j] = 0;
                else graph[i][j] = INF;
            }
        }
        //모든 경로 입력받기
        for(int i = 0; i < m; i++){
            int a = sc.nextInt();
            int b = sc.nextInt();
            //두 회사를 잇는 비용은 1, 양방향
            graph[a][b] = 1;
            graph[b][a] = 1;
        }

        //목적지와 경유지 입력받기
        int x = sc.nextInt();
        int y = sc.nextInt();

        //플로이드 워셜 알고리즘 수행
        for(int k = 1; k <= n; k++){
            for(int a = 1; a <= n; a++){
                for(int b = 1; b <= n; b++){
                    graph[a][b] = Math.min(graph[a][b], graph[a][k] + graph[k][b]);
                }
            }
        }

        int minDistance = graph[1][y] + graph[y][x];
        if(minDistance >= INF) System.out.println(-1);
        else System.out.println(minDistance);
    }
}

File: src/main/kotlin/shortestpath/미래_도시.kt
package shortestpath

import java.util.*
import kotlin.math.*

private const val INF = 1e9.toInt()

fun main(){
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    val m = sc.nextInt()

    val graph = mutableListOf<MutableList<Int>>()
    for(i in 0..n){
        graph.add(mutableListOf())
        for(j in 0..n){
            if(i == j) graph[i].add(0)
            else graph[i].add(INF)
        }
    }
    //경로 입력받기
    for(i in 0 until m){
        val a = sc.nextInt()
        val b = sc.nextInt()

        graph[a][b] = 1
        graph[b][a] = 1
    }

    val x = sc.nextInt()    //목적지
    val k = sc.nextInt()    //경유지

    //플로이드 워셜 알고리즘 수행
    for(l in 1..n){
        for(i in 1..n){
            for(j in 1..n){
                graph[i][j] = min(graph[i][j], graph[i][l] + graph[l][j])
            }
        }
    }

    when{
        graph[1][k] + graph[k][x] >= INF -> println(-1)
        else -> println(graph[1][k] + graph[k][x])
    }
}